File: src/UniversityStudent.java
import java.util.*;

/**
 * Concrete implementation of a Student representing a university student
 * in the Longhorn Network. This class extends the abstract Student class
 * and implements the connection strength calculation algorithm.
 * 
 * @author LonghornNetwork Team
 * @version 1.0
 */
public class UniversityStudent extends Student {
    /** The current roommate of this student */
    private UniversityStudent roommate;
    
    /** List of friends (other students) */
    private List<UniversityStudent> friends;
    
    /** Chat history with other students, mapped by student to list of messages */
    private Map<UniversityStudent, List<String>> chatHistory;
    
    /**
     * Constructs a new UniversityStudent with the specified attributes.
     * 
     * @param name The name of the student
     * @param age The age of the student
     * @param gender The gender of the student
     * @param year The academic year (1-4)
     * @param major The major field of study
     * @param gpa The grade point average
     * @param roommatePreferences List of preferred roommate names in order of preference
     * @param previousInternships List of previous internship company names
     */
    public UniversityStudent(String name, int age, String gender, int year, 
                            String major, double gpa, 
                            List<String> roommatePreferences, 
                            List<String> previousInternships) {
        this.name = name;
        this.age = age;
        this.gender = gender;
        this.year = year;
        this.major = major;
        this.gpa = gpa;
        this.roommatePreferences = new ArrayList<>(roommatePreferences);
        this.previousInternships = new ArrayList<>(previousInternships);
        this.roommate = null;
        this.friends = new ArrayList<>();
        this.chatHistory = new HashMap<>();
    }

    /**
     * Calculates the connection strength between this student and another student.
     * The connection strength is calculated based on the following criteria:
     * - Roommate: Add 4 if they are roommates
     * - Shared Internships: Add 3 for each shared internship
     * - Same Major: Add 2 if they share the same major
     * - Same Age: Add 1 if they are the same age
     * 
     * @param other The other student to calculate connection strength with
     * @return An integer representing the connection strength (0 or higher)
     */
    @Override
    public int calculateConnectionStrength(Student other) {
        if (!(other instanceof UniversityStudent)) {
            return 0;
        }
        
        UniversityStudent otherStudent = (UniversityStudent) other;
        int strength = 0;
        
        // Roommate: Add 4 if they are roommates
        if (this.roommate != null && this.roommate.equals(otherStudent)) {
            strength += 4;
        }
        
        // Shared Internships: Add 3 for each shared internship
        for (String internship : this.previousInternships) {
            if (otherStudent.previousInternships.contains(internship) && 
                !internship.equals("None") && !internship.isEmpty()) {
                strength += 3;
            }
        }
        
        // Same Major: Add 2 if they share the same major
        if (this.major != null && this.major.equals(otherStudent.major)) {
            strength += 2;
        }
        
        // Same Age: Add 1 if they are the same age
        if (this.age == otherStudent.age) {
            strength += 1;
        }
        
        return strength;
    }

    /**
     * Gets the current roommate of this student.
     * 
     * @return The UniversityStudent who is the roommate, or null if no roommate is assigned
     */
    public UniversityStudent getRoommate() {
        return roommate;
    }
    
    /**
     * Sets the roommate for this student.
     * 
     * @param roommate The student to set as roommate
     */
    public void setRoommate(UniversityStudent roommate) {
        this.roommate = roommate;
    }
    
    /**
     * Gets the list of friends.
     * 
     * @return The list of friends
     */
    public List<UniversityStudent> getFriends() {
        return friends;
    }
    
    /**
     * Adds a friend to this student's friend list.
     * 
     * @param friend The student to add as a friend
     */
    public void addFriend(UniversityStudent friend) {
        if (!friends.contains(friend)) {
            friends.add(friend);
        }
    }
    
    /**
     * Gets the chat history with a specific student.
     * 
     * @param student The student to get chat history with
     * @return The list of messages, or null if no chat history exists
     */
    public List<String> getChatHistory(UniversityStudent student) {
        return chatHistory.get(student);
    }
    
    /**
     * Adds a message to the chat history with a specific student.
     * 
     * @param student The student to add message to chat history with
     * @param message The message to add
     */
    public void addChatMessage(UniversityStudent student, String message) {
        chatHistory.putIfAbsent(student, new ArrayList<>());
        chatHistory.get(student).add(message);
    }

    /**
     * Returns a string representation of this UniversityStudent.
     * 
     * @return A string containing the student's information
     */
    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("Name: ").append(name).append("\n");
        sb.append("Age: ").append(age).append("\n");
        sb.append("Gender: ").append(gender).append("\n");
        sb.append("Year: ").append(year).append("\n");
        sb.append("Major: ").append(major).append("\n");
        sb.append("GPA: ").append(gpa).append("\n");
        sb.append("RoommatePreferences: ");
        if (roommatePreferences.isEmpty()) {
            sb.append("None");
        } else {
            sb.append(String.join(", ", roommatePreferences));
        }
        sb.append("\n");
        sb.append("PreviousInternships: ");
        if (previousInternships.isEmpty()) {
            sb.append("None");
        } else {
            sb.append(String.join(", ", previousInternships));
        }
        return sb.toString();
    }
    
    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || getClass() != obj.getClass()) return false;
        UniversityStudent that = (UniversityStudent) obj;
        return name != null && name.equals(that.name);
    }
    
    @Override
    public int hashCode() {
        return name != null ? name.hashCode() : 0;
    }
}

File: longhorn-network-react/src/models/UniversityStudent.ts
/**
 * UniversityStudent.ts
 *
 * Concrete implementation of a Student representing a university student
 * in the Longhorn Network. This class extends the abstract Student class
 * and implements the connection strength calculation algorithm.
 *
 * Key Responsibilities:
 * - Store and manage student information
 * - Calculate connection strengths with other students
 * - Manage roommate assignments
 * - Track friends and chat history
 *
 * Connection Strength Formula:
 * - Roommate: +4 points
 * - Each Shared Internship: +3 points
 * - Same Major: +2 points
 * - Same Age: +1 point
 *
 * @author LonghornNetwork Team
 * @version 1.0
 */

import { Student } from './Student';

/**
 * UniversityStudent class represents a student at the university
 * with full social networking capabilities including roommate matching,
 * friendship management, and chat functionality.
 */
export class UniversityStudent extends Student {
  /** The current roommate of this student (null if no roommate assigned) */
  private roommate: UniversityStudent | null = null;

  /** List of friends (other students) */
  private friends: UniversityStudent[] = [];

  /**
   * Chat history with other students.
   * Map structure: { student: [messages] }
   * Each entry maps a student to an array of messages exchanged with them.
   */
  private chatHistory: Map<UniversityStudent, string[]> = new Map();

  /**
   * Constructs a new UniversityStudent with the specified attributes.
   *
   * Usage Example:
   * ```typescript
   * const student = new UniversityStudent(
   *   "Alice",
   *   20,
   *   "Female",
   *   2,
   *   "Computer Science",
   *   3.8,
   *   ["Bob", "Charlie"],
   *   ["Google", "Microsoft"]
   * );
   * ```
   *
   * @param name - The name of the student (must be unique across all students)
   * @param age - The age of the student (typically 18-25)
   * @param gender - The gender of the student
   * @param year - The academic year (1=Freshman, 2=Sophomore, 3=Junior, 4=Senior)
   * @param major - The major field of study (e.g., "Computer Science", "Biology")
   * @param gpa - The grade point average on a 4.0 scale
   * @param roommatePreferences - Array of preferred roommate names in priority order
   * @param previousInternships - Array of company names where the student interned
   */
  constructor(
    name: string,
    age: number,
    gender: string,
    year: number,
    major: string,
    gpa: number,
    roommatePreferences: string[],
    previousInternships: string[]
  ) {
    super(name, age, gender, year, major, gpa, roommatePreferences, previousInternships);
  }

  /**
   * Calculates the connection strength between this student and another student.
   *
   * The connection strength is calculated using the following formula:
   * - Roommate: Add 4 if they are currently roommates
   * - Shared Internships: Add 3 for each internship they both completed
   * - Same Major: Add 2 if they share the same major
   * - Same Age: Add 1 if they are the same age
   *
   * Algorithm Complexity: O(n*m) where n and m are the number of internships
   *
   * Edge Cases:
   * - Returns 0 if no connections exist
   * - Ignores "None" or empty internship entries
   * - Returns 0 if the other parameter is not a UniversityStudent
   *
   * Example:
   * ```typescript
   * const alice = new UniversityStudent(...);
   * const bob = new UniversityStudent(...);
   * alice.setRoommate(bob); // They are roommates
   * // If they also share the same major and age:
   * const strength = alice.calculateConnectionStrength(bob); // Returns 7 (4+2+1)
   * ```
   *
   * @param other - The other student to calculate connection strength with
   * @returns An integer representing the connection strength (0 or higher)
   */
  public calculateConnectionStrength(other: Student): number {
    if (!(other instanceof UniversityStudent)) {
      return 0;
    }

    let strength = 0;

    // Roommate: Add 4 if they are roommates
    if (this.roommate !== null && this.roommate === other) {
      strength += 4;
    }

    // Shared Internships: Add 3 for each shared internship
    // Filter out "None" and empty values before comparison
    const validInternships = this.previousInternships.filter(
      (internship) => internship && internship !== 'None' && internship.trim() !== ''
    );

    for (const internship of validInternships) {
      if (other.previousInternships.includes(internship)) {
        strength += 3;
      }
    }

    // Same Major: Add 2 if they share the same major
    if (this.major && this.major === other.major) {
      strength += 2;
    }

    // Same Age: Add 1 if they are the same age
    if (this.age === other.age) {
      strength += 1;
    }

    return strength;
  }

  /**
   * Gets the current roommate of this student.
   *
   * @returns The UniversityStudent who is the roommate, or null if no roommate is assigned
   */
  public getRoommate(): UniversityStudent | null {
    return this.roommate;
  }

  /**
   * Sets the roommate for this student.
   *
   * Note: This method only updates this student's roommate reference.
   * To create a bidirectional roommate relationship, call setRoommate on both students.
   *
   * Example:
   * ```typescript
   * alice.setRoommate(bob);
   * bob.setRoommate(alice);
   * ```
   *
   * @param roommate - The student to set as roommate, or null to clear roommate
   */
  public setRoommate(roommate: UniversityStudent | null): void {
    this.roommate = roommate;
  }

  /**
   * Gets the list of friends.
   *
   * @returns A copy of the friends array to prevent external modification
   */
  public getFriends(): UniversityStudent[] {
    return [...this.friends];
  }

  /**
   * Adds a friend to this student's friend list.
   *
   * Thread-Safety Note: In a multi-threaded environment (simulated with async operations),
   * this method should be called within synchronized blocks.
   *
   * @param friend - The student to add as a friend
   */
  public addFriend(friend: UniversityStudent): void {
    if (!this.friends.includes(friend)) {
      this.friends.push(friend);
    }
  }

  /**
   * Gets the chat history with a specific student.
   *
   * @param student - The student to get chat history with
   * @returns The array of messages exchanged with the student, or undefined if no history exists
   */
  public getChatHistory(student: UniversityStudent): string[] | undefined {
    return this.chatHistory.get(student);
  }

  /**
   * Gets all chat histories for this student.
   *
   * @returns A Map of students to their chat message arrays
   */
  public getAllChatHistories(): Map<UniversityStudent, string[]> {
    return new Map(this.chatHistory);
  }

  /**
   * Adds a message to the chat history with a specific student.
   *
   * Thread-Safety Note: In a multi-threaded environment (simulated with async operations),
   * this method should be called within synchronized blocks.
   *
   * @param student - The student to add message to chat history with
   * @param message - The message content to add
   */
  public addChatMessage(student: UniversityStudent, message: string): void {
    if (!this.chatHistory.has(student)) {
      this.chatHistory.set(student, []);
    }
    this.chatHistory.get(student)!.push(message);
  }

  /**
   * Gets the name of this student.
   *
   * @returns The student's name
   */
  public getName(): string {
    return this.name;
  }

  /**
   * Gets the age of this student.
   *
   * @returns The student's age
   */
  public getAge(): number {
    return this.age;
  }

  /**
   * Gets the gender of this student.
   *
   * @returns The student's gender
   */
  public getGender(): string {
    return this.gender;
  }

  /**
   * Gets the academic year of this student.
   *
   * @returns The student's year (1-4)
   */
  public getYear(): number {
    return this.year;
  }

  /**
   * Gets the major of this student.
   *
   * @returns The student's major
   */
  public getMajor(): string {
    return this.major;
  }

  /**
   * Gets the GPA of this student.
   *
   * @returns The student's GPA
   */
  public getGPA(): number {
    return this.gpa;
  }

  /**
   * Gets the roommate preferences of this student.
   *
   * @returns A copy of the roommate preferences array
   */
  public getRoommatePreferences(): string[] {
    return [...this.roommatePreferences];
  }

  /**
   * Gets the previous internships of this student.
   *
   * @returns A copy of the previous internships array
   */
  public getPreviousInternships(): string[] {
    return [...this.previousInternships];
  }

  /**
   * Returns a string representation of this UniversityStudent.
   * Useful for debugging and logging.
   *
   * @returns A formatted string containing the student's information
   */
  public toString(): string {
    const lines = [
      `Name: ${this.name}`,
      `Age: ${this.age}`,
      `Gender: ${this.gender}`,
      `Year: ${this.year}`,
      `Major: ${this.major}`,
      `GPA: ${this.gpa}`,
      `RoommatePreferences: ${this.roommatePreferences.length > 0 ? this.roommatePreferences.join(', ') : 'None'}`,
      `PreviousInternships: ${this.previousInternships.length > 0 ? this.previousInternships.join(', ') : 'None'}`
    ];
    return lines.join('\n');
  }

  /**
   * Checks equality based on student name.
   * Two students are considered equal if they have the same name.
   *
   * @param other - The other student to compare with
   * @returns true if the students have the same name, false otherwise
   */
  public equals(other: UniversityStudent): boolean {
    return this.name === other.name;
  }
}
